shared-libs/infodoc/src/infodoc.js

```diff
@@ -0,0 +1,288 @@
+  };
+};
+
+const findInfoDocs = (database, ids) => {
+  return database
+    .allDocs({ keys: ids, include_docs: true })
+    .then(results => results.rows);
+};
+
+//
+// Given a set of changes, find all the infoDocs or create them as necessary. Also takes care of
+// migrating legacy infodocs from the medic db, and legacy transition information from records.
+//
+// @param      {Array}  changes  an array of PouchDB changes objects
+// @return     {Array}  array of infodocs. NB: will not necessarily be in the same order as the
+//                      changes were passed in
+//
+const resolveInfoDocs = (changes, writeDirtyInfoDocs) => {
+  const splitInfoDocRows = results => {
+    return results.reduce((acc, row) => {
+      if (!row.doc) {
+        acc.missing.push({_id: row.key});
+      } else if (!row.doc.transitions) {
+        // No transitions may mean that API created this infodoc on write but sentinel hasn't seen
+        // it yet. It's possible that there is a legacy infodoc with transition information.
+        acc.missingTransitions.push(row.doc);
+      } else {
+        acc.valid.push(row.doc);
+      }
+
+      return acc;
+
+    }, {valid: [], missing: [], missingTransitions: []});
+  };
+
+  if (!changes || !changes.length) {
+    return Promise.resolve();
+  }
+
+  const infoDocIds = changes.map(change => getInfoDocId(change.id));
+
+  // First attempt, directly from sentinel where they should live
+  // @4.0 remove all of this and replace it with just a simple get (+ a migration to resolve legacy issues)
+  return findInfoDocs(db.sentinel, infoDocIds)
+    .then(results => {
+      const { valid, missing, missingTransitions: missingTransitionsSentinel } = splitInfoDocRows(results);
+
+      const lookForInMedic = missing.concat(missingTransitionsSentinel).map(r => r._id);
+
+      if (!lookForInMedic.length) {
+        return valid;
+      }
+
+      // the infodocs missing transitions are still valid. We distinguish between them so we can
+      // check for medic-db infodocs and if they exist transfer the transition data over
+      const infoDocs = valid.concat(missingTransitionsSentinel);
+
+      // Second attempt, look for old infodocs in the Medic DB.
+      return findInfoDocs(db.medic, lookForInMedic)
+        .then(results => {
+          const migratedInfoDocs = [];
+          const { valid, missing, missingTransitions: missingTransitionsMedic } = splitInfoDocRows(results);
+
+          // There is no interesting reason for a legacy medic infodoc to not have transitions, it's valid enough!
+          valid.push(...missingTransitionsMedic);
+
+          // Convert valid MedicDB infodocs into Sentinel ones
+          valid.forEach(medicInfoDoc => {
+            const sentinelInfoDoc = missingTransitionsSentinel.find(d => d._id === medicInfoDoc._id);
+            const doc = changes.find(change => change.doc && change.doc._id === medicInfoDoc._id) || {};
```

Comment:
Sorry I missed it before, but there still are some issues here. 
You're comparing `change.doc._id === medicInfoDoc._id` where `change.doc._id` will be "some-uuid" and `medicInfoDoc._id` will be "some-uuid-info", the change will never be found. 
Also, the `changes.find` would return the `change` not the doc, and you're expecting the doc a few lines lower. 

```
const change = changes.find(change => change.doc && getInfoDocId(change.doc._id) === medicInfoDoc._id) || {};
const doc = change.doc || {};
```

I think a unit test to cover this is needed, to check if transitions are actually copied correctly from the medic doc to the sentinel info doc.